alf/algorithms/trac_algorithm.py

```diff
@@ -0,0 +1,171 @@
+            debug_summaries=debug_summaries,
+            name=name)
+        action_param_spec = nest_map(lambda spec: spec.input_params_spec,
+                                     actor_network.output_spec)
+        spec = TracExperience(
+            observation=observation_spec,
+            step_type=tf.TensorSpec((), tf.int32),
+            state=self.train_state_spec,
+            action_param=action_param_spec)
+        self._buffer = DataBuffer(
+            spec, capacity=(buffer_size + 1) * batch_size)
+        self._trusted_updater = None
+        self._kl_clips = nest_map(lambda _: kl_clip, self.action_spec)
+        self._batch_size = batch_size
+
+    def rollout(self, time_step: ActionTimeStep, state: ActorCriticState):
+        policy_step = super().rollout(time_step, state)
+        action_param = common.get_distribution_params(policy_step.action)
+        exp = TracExperience(
+            observation=time_step.observation,
+            step_type=time_step.step_type,
+            state=state,
+            action_param=action_param)
+
+        # run_if() is unnecessary. But TF has error if run_if is removed.
+        # TODO: find the minimum condition to reproduce this bug.
+        run_if(
+            tf.shape(time_step.step_type)[0] >
+            0, lambda: self._buffer.add_batch(exp))
+
+        if self._trusted_updater is None:
+            self._trusted_updater = TrustedUpdater(
+                self._actor_network.trainable_variables)
+        return policy_step
+
+    def after_train(self):
+        """Adjust actor parameter according to KL-divergence."""
+        kls, steps = self._trusted_updater.adjust_step(self._calc_kl,
+                                                       self._kl_clips)
+
+        def _summarize():
+            with self.name_scope:
+                for i, kl in enumerate(tf.nest.flatten(kls)):
+                    tf.summary.scalar("action_kl/%s" % i, kl)
+                tf.summary.scalar("adjust_steps", steps)
+
+        common.run_if(common.should_record_summaries(), _summarize)
+        super().after_train()
+
+    @tf.function
+    def _calc_kl(self):
+        """Calculate the symmetrical KL-devengence between old/new action dist.
+
+        Equation: KL(d1||d2) + KL(d2||d1)
+        """
+        # Remove the samples from the last step from the buffer because it may
+        # repeated if the final_step_mode of OnPolicyDriver is FINAL_STEP_REDO.
+        # See comment of OnPolicyDriver for detail.
+        self._buffer.pop(self._batch_size)
+        size = self._buffer.current_size
+        total_kls = nest_map(lambda _: tf.zeros(()), self.action_spec)
+        for t in tf.range(0, size, self._batch_size):
+            exp = self._buffer.get_batch_by_indices(
+                tf.range(t, t + self._batch_size))
+            # Here we use stored states to calculate action. It might be better
+            # to use the states calculated using the new weights.
```

Comment:
Might have problems with RNN-actor networks.

Reply:
It's not clear which is good for RNN. It's relatively easy to use the RNN state if we want.